src/hotspot/share/opto/vectorization.hpp

```diff
@@ -439,6 +443,109 @@ class VLoopTypes : public StackObj {
+class VLoopDependencyGraph : public StackObj {
+private:
+  class DependencyNode;
+
+  Arena*                   _arena;
+  const VLoop&             _vloop;
+  const VLoopBody&         _body;
+  const VLoopMemorySlices& _memory_slices;
+
+  // bb_idx -> DependenceNode*
+  GrowableArray<DependencyNode*> _dependency_nodes;
+
+  // Node depth in DAG: bb_idx -> depth
+  GrowableArray<int> _depth;
+
+protected:
+  const VLoopBody& body() const { return _body; }
+
+public:
+  VLoopDependencyGraph(Arena* arena,
+                       const VLoop& vloop,
+                       const VLoopBody& body,
+                       const VLoopMemorySlices& memory_slices) :
+    _arena(arena),
+    _vloop(vloop),
+    _body(body),
+    _memory_slices(memory_slices),
+    _dependency_nodes(arena,
+                      vloop.estimated_body_length(),
+                      vloop.estimated_body_length(),
+                      nullptr),
+    _depth(arena,
+           vloop.estimated_body_length(),
+           vloop.estimated_body_length(),
+           0) {}
+  NONCOPYABLE(VLoopDependencyGraph);
+
+  void construct();
+  bool independent(Node* s1, Node* s2) const;
+  bool mutually_independent(const Node_List* nodes) const;
+
+private:
+  void add_node(MemNode* n, GrowableArray<int>& memory_pred_edges);
+  int depth(const Node* n) const { return _depth.at(_body.bb_idx(n)); }
+  void set_depth(const Node* n, int d) { _depth.at_put(_body.bb_idx(n), d); }
+  void compute_depth();
+  NOT_PRODUCT( void print() const; )
+
+  const DependencyNode* dependency_node(const Node* n) const {
+    return _dependency_nodes.at(_body.bb_idx(n));
+  }
+
+  class DependencyNode : public ArenaObj {
+  private:
+    MemNode* _node; // Corresponding ideal node
+    const uint _memory_pred_edges_length;
+    int* _memory_pred_edges; // memory pred-edges, mapping to bb_idx
+  public:
+    DependencyNode(MemNode* n, GrowableArray<int>& memory_pred_edges, Arena* arena);
+    NONCOPYABLE(DependencyNode);
+    uint memory_pred_edges_length() const { return _memory_pred_edges_length; }
+
+    int memory_pred_edge(uint i) const {
+      assert(i < _memory_pred_edges_length, "bounds check");
+      return _memory_pred_edges[i];
+    }
+  };
+
+public:
+  // Iterator for dependency graph predecessors of a node.
+  class PredsIterator : public StackObj {
+  private:
+    const VLoopDependencyGraph& _dependency_graph;
+
+    const Node* _node;
+    const DependencyNode* _dependency_node;
+
+    Node* _current;
+
+    // Iterate in node->in(i)
+    int _next_pred;
+    int _end_pred;
+
+    // Iterate in dependency_node->memory_pred_edge(i)
+    int _next_memory_pred;
+    int _end_memory_pred;
+  public:
+    PredsIterator(const VLoopDependencyGraph& dependency_graph, const Node* node);
+    NONCOPYABLE(PredsIterator);
+    void next();
+    bool done() const { return _current == nullptr; }
+    Node* current() const { assert(!done(), "not done yet"); return _current; }
```

Comment:
For two statements, I suggest to go with multiple lines:
```suggestion
    Node* current() const { 
      assert(!done(), "not done yet"); 
      return _current;
    }
```